code/autoupdater/autoupdater.c

```diff
@@ -0,0 +1,780 @@
+    int64_t len;
+    int update;
+    int rollback;
+    struct ManifestItem *next;
+} ManifestItem;
+
+static ManifestItem *manifest = NULL;
+
+static void freeManifest(void)
+{
+    ManifestItem *item = manifest;
+    manifest = NULL;
+
+    while (item != NULL) {
+        ManifestItem *next = item->next;
+        free(item->fname);
+        free(item);
+        item = next;
+    }
+    manifest = NULL;
+}
+
+static const char *timestamp(void)
+{
+    time_t t = time(NULL);
+    char *retval = asctime(localtime(&t));
+    if (retval) {
+        char *ptr;
+        for (ptr = retval; *ptr; ptr++) {
+            if ((*ptr == '\r') || (*ptr == '\n')) {
+                *ptr = '\0';
+                break;
+            }
+        }
+    }
+    return retval ? retval : "[date unknown]";
+}
+
+
+static FILE *logfile = NULL;
+
+#define SDL_PRINTF_VARARG_FUNC( fmtargnumber ) __attribute__ (( format( __printf__, fmtargnumber, fmtargnumber+1 )))
+
+static void info(const char *str)
+{
+    fputs(str, logfile);
+    fputs("\n", logfile);
+    fflush(logfile);
+}
+
+static void infof(const char *fmt, ...)
+#if defined(__GNUC__) || defined(__clang__)
+__attribute__ (( format( __printf__, 1, 2 )))
+#endif
+;
+
+static void infof(const char *fmt, ...)
+{
+    va_list ap;
+    va_start(ap, fmt);
+    vfprintf(logfile, fmt, ap);
+    va_end(ap);
+    fputs("\n", logfile);
+    fflush(logfile);
+}
+
+static void restoreRollbacks(void)
+{
+    /* you can't call die() in this function! If this fails, you're doomed. */
+    ManifestItem *item;
+    for (item = manifest; item != NULL; item = item->next) {
+        if (item->rollback) {
+            char rollbackPath[64];
+            snprintf(rollbackPath, sizeof (rollbackPath), "updates/rollbacks/%d", item->rollback);
+            infof("restore rollback: '%s' -> '%s'", rollbackPath, item->fname);
+            remove(item->fname);
+            rename(rollbackPath, item->fname);
+        }
+    }
+}
+
+static void die(const char *why) NEVER_RETURNS;
+
+
+#ifndef _WIN32  /* hooray for Unix linker hostility! */
+#undef curl_easy_setopt
+#include <dlfcn.h>
+typedef void (*CURLFN_curl_easy_cleanup)(CURL *curl);
+typedef CURL *(*CURLFN_curl_easy_init)(void);
+typedef CURLcode (*CURLFN_curl_easy_setopt)(CURL *curl, CURLoption option, ...);
+typedef CURLcode (*CURLFN_curl_easy_perform)(CURL *curl);
+typedef CURLcode (*CURLFN_curl_global_init)(long flags);
+typedef void (*CURLFN_curl_global_cleanup)(void);
+
+static CURLFN_curl_easy_cleanup CURL_curl_easy_cleanup;
+static CURLFN_curl_easy_init CURL_curl_easy_init;
+static CURLFN_curl_easy_setopt CURL_curl_easy_setopt;
+static CURLFN_curl_easy_perform CURL_curl_easy_perform;
+static CURLFN_curl_global_init CURL_curl_global_init;
+static CURLFN_curl_global_cleanup CURL_curl_global_cleanup;
+
+static void load_libcurl(void)
+{
+    #ifdef __APPLE__
+    const char *libname = "libcurl.4.dylib";
+    #else
+    const char *libname = "libcurl.so.4";
```

Comment:
True, our ports system builds ioquake3 from source and links against proper libcurl:

```
$ ldd `which ioquake3` 
/usr/local/bin/ioquake3:
        Start            End              Type Open Ref GrpRef Name
        000018a312100000 000018a313ef1000 exe  2    0   0      /usr/local/bin/ioquake3
        000018a513eb5000 000018a5141c0000 rlib 0    1   0      /usr/local/lib/libSDL2.so.0.3
        000018a5d770b000 000018a5d7978000 rlib 0    1   0      /usr/local/lib/libcurl.so.25.9
        000018a55b977000 000018a55bd7d000 rlib 0    1   0      /usr/local/lib/libogg.so.6.2
        000018a5c5d13000 000018a5c5f3a000 rlib 0    2   0      /usr/lib/libm.so.10.0
        000018a536fa6000 000018a5371b5000 rlib 0    2   0      /usr/lib/libpthread.so.23.0
        000018a5b5149000 000018a5b5415000 rlib 0    1   0      /usr/lib/libc.so.89.3
        000018a5fc986000 000018a5fcb93000 rlib 0    1   0      /usr/lib/libsndio.so.6.1
        000018a556941000 000018a556b46000 rlib 0    1   0      /usr/lib/libusbhid.so.7.0
        000018a51b3ba000 000018a51b5df000 rlib 0    1   0      /usr/local/lib/libnghttp2.so.0.10
        000018a5fc286000 000018a5fc4db000 rlib 0    1   0      /usr/lib/libssl.so.43.2
        000018a5141c0000 000018a51458a000 rlib 0    2   0      /usr/lib/libcrypto.so.41.1
        000018a51ca55000 000018a51cc6b000 rlib 0    1   0      /usr/lib/libz.so.5.0
        000018a563f00000 000018a563f00000 rtld 0    1   0      /usr/libexec/ld.so
$ 
```

we would also never enable the auto updater on our platform so as long as the code can be compiled without it then we are fine.